modules/dag-l0/src/main/scala/org/tessellation/dag/l0/domain/snapshot/programs/GlobalSnapshotEventCutter.scala

```diff
@@ -0,0 +1,144 @@
+      }
+
+      val ordering: Ordering[PeerIdEvent] = Ordering.by[PeerIdEvent, (Long, Hash)](e => (e.fee, e.hash))
+
+      case class PeerIdStateChannelEvent(id: PeerId, event: StateChannelEvent) extends PeerIdEvent {
+        val fee: Long = event.snapshotBinary.value.fee.value.value
+        def hash: Hash = event.snapshotBinary.lastSnapshotHash
+        def toGlobalSnapshotEvent: GlobalSnapshotEvent = event.asLeft[DAGEvent]
+      }
+
+      case class PeerIdHashedEvent(id: PeerId, event: Hashed[Block]) extends PeerIdEvent {
+        val fee: Long = event.signed.value.transactions.toIterable.map(_.value.fee.value.value).sum
+        def hash: Hash = event.hash
+        def toGlobalSnapshotEvent: GlobalSnapshotEvent = event.signed.asRight[StateChannelEvent]
+
+        def isMyChild(that: PeerIdHashedEvent): Boolean =
+          that.event.signed.value.parent.contains_(event.ownReference)
+      }
+
+      class OrderedEvents(sortedStateEvents: List[PeerIdStateChannelEvent], hashedEvents: List[PeerIdHashedEvent]) {
+        def next: Option[(PeerIdEvent, OrderedEvents)] = {
+          val (parents, nonParents) =
+            hashedEvents.foldLeft((List.empty[PeerIdHashedEvent], List.empty[PeerIdHashedEvent])) {
+              case ((parents, nonParents), p) if hashedEvents.exists(p.isMyChild) => (p :: parents, nonParents)
+              case ((parents, nonParents), p)                                     => (parents, p :: nonParents)
+            }
+
+          (sortedStateEvents, nonParents.sorted(ordering)) match {
+            case (Nil, Nil) =>
+              none
+            case (h :: t, Nil) =>
+              (h, new OrderedEvents(t, Nil)).some
+            case (Nil, h :: t) =>
+              (h, new OrderedEvents(Nil, t ++ parents)).some
+            case (h1 :: t1, h2 :: t2) =>
+              if (h1.fee < h2.fee || (h1.fee == h2.fee && sortedStateEvents.length >= hashedEvents.length))
+                (h1, new OrderedEvents(t1, hashedEvents)).some
+              else
+                (h2, new OrderedEvents(sortedStateEvents, t2 ++ parents)).some
+          }
+        }
+      }
+
+      object OrderedEvents {
+        // NOTE: Order for removal: NOT parent → Fee → Split by event type → Hash
+        def apply(events: GlobalPeerIdEventMap): F[OrderedEvents] =
+          unpackGlobalPeerIdEvents(events).map {
+            case (stateChannelEvents, hashedEvents) =>
+              new OrderedEvents(stateChannelEvents.sorted(ordering), hashedEvents)
+          }
+
+        def unpackGlobalPeerIdEvents(events: GlobalPeerIdEventMap): F[(List[PeerIdStateChannelEvent], List[PeerIdHashedEvent])] =
+          events.view
+            .mapValues(es => unpackEither(es.map(_._2)))
+            .toList
+            .traverse {
+              case (peerId, (stateChannelEvents, dagEvents)) =>
+                dagEvents.traverse(_.toHashed[F]).map(hashedEvents => peerId -> (stateChannelEvents, hashedEvents))
```

Comment:
Could you perform the mapping to `PeerIdStateChannelEvent` and `PeerIdHashedEvent` here? And after it just flatten it or add up using a predefined `Monoid`? I think that could work 🤔  There would also be no need to pass the `peerId` and the `foldLeft`  part below aggregating the events won't be necessary. It looks a bit boilerplate'y right now.

Reply:
Something like:
```
            .traverse {
              case (peerId, (stateChannelEvents, dagEvents)) =>
                dagEvents.traverse(_.toHashed[F]).map { hashedEvents =>
                  val scEvents = stateChannelEvents.map(PeerIdStateChannelEvent(peerId, _))
                  val dagEvents = hashedEvents.map(PeerIdHashedEvent(peerId, _))

                  (scEvents, dagEvents)
                }
            }
```
and then the part adding the matching lists together (didn't try to come up with this part yet)

Reply:
I guess this:
```
.map(_.combineAll)
```

Reply:
Check if this implementation is the right one:
```
        def unpackGlobalPeerIdEvents(events: GlobalPeerIdEventMap): F[(List[PeerIdStateChannelEvent], List[PeerIdHashedEvent])] =
          events.view
            .mapValues(_.map(_._2).partitionMap(identity()))
            .toList
            .traverse {
              case (peerId, (stateChannelEvents, dagEvents)) =>
                dagEvents.traverse(_.toHashed[F]).map { hashedEvents =>
                  val scEvents = stateChannelEvents.map(PeerIdStateChannelEvent(peerId, _))
                  val dagEvents = hashedEvents.map(PeerIdHashedEvent(peerId, _))

                  (scEvents, dagEvents)
                }
            }
            .map(_.combineAll)
```

Reply:
@TheMMaciek That's much cleaner, thank you.